### PR TITLE
Feature/revisit dynamic pathing

### PR DESCRIPTION
**Pull request by JohnGilb**
# Why are we doing this?
The existing pathing logic was a bit old and not well documented; updated and fixed several bugs.
# Whats changing?
* Lots of bugfixes related to interesting collision types, particularly when the ray would cross opposite sides of the obstacle.
* If the robot gets really disconnected from the ghost (collision, possibly human intervention), it can recover and get back on the line of travel.
# Questions/notes for reviewers

# How this was tested
- [x] unit tests added
- [x] tested on robot
- [x] tested in local simulator

### CODE REVIEW

File: src/main/java/xbot/common/command/BaseMaintainerCommand.java [hunk 1]
```diff
@@ -207,11 +207,6 @@ protected boolean getErrorWithinTolerance() {
      */
     protected abstract double getHumanInputMagnitude();
 
-    @Override
-    public String getPrefix() {
-        return subsystemToMaintain.getPrefix() + getName() + "/";
```
Comment (JohnGilb) [inline src/main/java/xbot/common/command/BaseMaintainerCommand.java:212]: This causes a nullref as the subsystem can't easily be created. Maintainer records will need to live under their own command, which at least matches up with other commands.

File: src/main/java/xbot/common/trajectory/LowResField.java [hunk 2]
```diff
@@ -24,13 +24,306 @@ public LowResField() {
...
+        Translation2d currentSource = currentPose.getTranslation();
+        //      b) initialize an empty list of swerve points.
+        List<XbotSwervePoint> path = new ArrayList<XbotSwervePoint>();
+        // 2) Create a Stack of XbotSwervePoints, which will be used to help us deal with multiple intermediate points.
+        Stack<XbotSwervePoint> swervePointStack = new Stack<XbotSwervePoint>();
+
+        XbotSwervePoint specialFinalPoint = null;
+
+        for (Obstacle o : obstacles) {
+            o.resetCorners();
```
Comment (aschokking) [inline src/main/java/xbot/common/trajectory/LowResField.java:107]: it feels a little icky to me for this function to have side effects on obstacles, maybe this should be working on a copy of them?

File: src/main/java/xbot/common/trajectory/LowResField.java [hunk 3]
```diff
@@ -24,13 +24,306 @@ public LowResField() {
...
+                    if (specialFinalPoint != null) {
+                        path.add(specialFinalPoint);
+                    }
+
+                    completed = true;
+                }
+            }
+
+        }
+        return path;
```
Comment (aschokking) [inline src/main/java/xbot/common/trajectory/LowResField.java:324]: this is a pretty enormous function (not just from the comments), what do you think about breaking it down into a few smaller pieces so the outer function reads a little easier?

I also wonder if those inner parts wouldn't be easier to unit test as well?

File: src/test/java/xbot/common/trajectory/LowResFieldTest.java [hunk 4]
```diff
@@ -57,6 +59,7 @@ private FieldPose flipFieldPose(FieldPose current) {
                 new WrappedRotation2d(180 - current.getHeading().getDegrees()));
     }
 
+    @Ignore
```
Comment (aschokking) [inline src/test/java/xbot/common/trajectory/LowResFieldTest.java:62]: ⭐ could you add a comment about why these are ignored? Is it a TODO to get them passing again?
eg:
`@Ignore("Broken from change X, TODO to get passing again.")`

**Review by aschokking — APPROVED** [review]
Wooh there's a lot here! I don't 100% follow all of the logic but have some readability suggestions

File: src/main/java/xbot/common/trajectory/LowResField.java [hunk 5]
```diff
@@ -24,13 +24,306 @@ public LowResField() {
...
 
     public void addObstacle(Obstacle o) {
-        obstacles.add(o);
+        Obstacle potentialObstacleToRemove = null;
+        for (Obstacle existingObstacle : obstacles) {
+            if (existingObstacle.intersects(o) || existingObstacle.contains(o) || o.contains(existingObstacle)) {
+                log.info("Collides with existing obstacle. Combining.");
+                existingObstacle.absorbObstacle(o);
+                potentialObstacleToRemove = existingObstacle;
+                break;
```
[the remark this answers is not in the document]
Comment (JohnGilb) [reply src/main/java/xbot/common/trajectory/LowResField.java:33]: Yes; let me add that as documentation.

File: src/main/java/xbot/common/trajectory/LowResField.java [hunk 6]
```diff
@@ -24,13 +24,306 @@ public LowResField() {
...
+    }
+
+    /**
+     * Generates a path from the current position to the target point, avoiding known obstacles.
+     * Very generally, it will recursively solve paths nearby the target, and then step foward to approach the target.
+     * @param currentPose Starting point
+     * @param targetPoint Terminating Point
+     * @return A List of XbotSwervePoints that can be followed to reach the destination without collision.
+     */
+    public List<XbotSwervePoint> generatePathFrontToBack(Pose2d currentPose, XbotSwervePoint targetPoint) {
```
[the remark this answers is not in the document]
Comment (JohnGilb) [reply src/main/java/xbot/common/trajectory/LowResField.java:64]: Good point - after some testing, I think the new method can take the mantle of `generatePath()` and we can retire the old one.

File: src/main/java/xbot/common/trajectory/LowResField.java [hunk 7]
```diff
@@ -24,13 +24,306 @@ public LowResField() {
...
+        // New attempt at a pathfinding algorithm that maybe a little more resilient. The original algorithm had
```
[the remark this answers is not in the document]
Comment (JohnGilb) [reply src/main/java/xbot/common/trajectory/LowResField.java:65]: original obliterated

**Pull request merged**